src/sync/sync_manager.cpp

```diff
@@ -0,0 +1,341 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "sync_manager.hpp"
+
+#include "impl/sync_client.hpp"
+#include "impl/sync_file.hpp"
+#include "impl/sync_metadata.hpp"
+#include "sync_session.hpp"
+#include "sync_user.hpp"
+
+#include <thread>
+
+using namespace realm;
+using namespace realm::_impl;
+
+SyncManager& SyncManager::shared()
+{
+    static SyncManager& manager = *new SyncManager;
+    return manager;
+}
+
+void SyncManager::configure_file_system(const std::string& base_file_path,
+                                        MetadataMode metadata_mode,
+                                        util::Optional<std::vector<char>> custom_encryption_key)
+{
+    std::vector<std::tuple<std::string, std::shared_ptr<SyncUser>>> users_to_add;
+    REALM_ASSERT(!m_file_manager && !m_metadata_manager);   // Can only call this once
```

Comment:
This doesn't work for react-native as you can refresh an app causing this to get called multiple times. We need to allow multiple calls to this and to cleanup appropriately when this occurs.